docs/how-to/hip_runtime_api/error_handling.rst

```diff
@@ -0,0 +1,212 @@
+runtime API call **without** resetting the error code:
+
+.. code-block:: cpp
+
+  hipError_t err = hipPeekAtLastError();
+  if (err != hipSuccess)
+  {
+      std::cerr << "HIP error: " << hipGetErrorString(err) << std::endl;
+  }
+
+:cpp:func:`hipGetErrorName` converts a HIP error code to a string representing
+the error name:
+
+.. code-block:: cpp
+
+  std::cerr << "Error name: " << hipGetErrorName(err) << std::endl;
+
+:cpp:func:`hipGetErrorString` converts a HIP error code to a string describing
+the error:
+
+.. code-block:: cpp
+
+  std::cerr << "Error description: " << hipGetErrorString(err) << std::endl;
+
+Best practices
+--------------------------------------------------------------------------------
+
+1. Check errors after each API call
+
+  Always check the return value of HIP API calls to catch errors early. For
+  example:
+
+  .. code-block:: cpp
+
+    hipError_t err = hipMalloc(&d_A, size);
+    if (err != hipSuccess) {
+        std::cerr << "hipMalloc failed: " << hipGetErrorString(err) << std::endl;
+        return -1;
+    }
+
+2. Use macros for error checking
+
+  Define macros to simplify error checking and reduce code duplication. For
+  example:
+
+  .. code-block:: cpp
+
+    #define HIP_CHECK(expression)                  \
+    {                                              \
+        const hipError_t status = expression;      \
+        if(status != hipSuccess){                  \
+            std::cerr << "HIP error "              \
+                      << status << ": "            \
+                      << hipGetErrorString(status) \
+                      << " at " << __FILE__ << ":" \
+                      << __LINE__ << std::endl;    \
+        }                                          \
+    }
+
+    // Allocate memory on the device with error check.
+    HIP_CHECK(hipMalloc(&device_pointer, N * sizeof(float);));
+
+3. Handle errors gracefully
+
+   Ensure the application can handle errors gracefully, such as by freeing
+   resources or providing meaningful error messages to the user.
+
+Complete example
+--------------------------------------------------------------------------------
+
+A complete example demonstrating error handling:
+
+.. code-block:: cpp
+
+  #include <hip/hip_runtime.h>
+  #include <iostream>
+
+  #define HIP_CHECK(expression)                  \
+  {                                              \
+      const hipError_t status = expression;      \
+      if(status != hipSuccess){                  \
+          std::cerr << "HIP error "              \
+                    << status << ": "            \
+                    << hipGetErrorString(status) \
+                    << " at " << __FILE__ << ":" \
+                    << __LINE__ << std::endl;    \
+      }                                          \
+  }
+
+  // Addition of two values.
+  __global__ void add(int *a, int *b, int *c) {
+      *c = *a + *b;
+  }
```

Comment:
Even if the focus here is on the error handling, I think we shouldn't encourage people to program kernels this way. With some slight changes, this could be a vector(array)-add-like kernel. Then we also wouldn't allocate memory for "scalar" variables, and wouldn't launch a kernel with 1 thread in 1 block

Reply:
I disagree with this. I do not think the 1 thread in 1 block kernel sample code is encouraging people to a bad habit.

Obviously no one should add two values at kernel side, but for showing the error handling where the kernel is not the main focus of the example code can help to understand the error handling faster.

I can add a comment, that this kind of kernel only is a simple example, which perform better at host side.

Reply:
I didn't mean the performance of this kernel, or whether it even is suited for GPUs, I wouldn't even mind if it was a only the initialization of an array, even those kernels can have their justifications on GPUs, if they in turn avoid a copy from host to device.

This code doesn't provide any parallelism, even if it was launched with more threads, as it does not use threadIds etc. Expanding this to properly use several threads wouldn't expand the code by too much, it mostly would be adding a "array_size" variable, that would be added to the malloc and memcpy calls, and would add two lines overall to the kernel - getting the threads index and checking the bounds of the array. I don't think that would distract the user from the error handling in this example.

Reply:
You got your size.